server/src-lib/Hasura/RQL/DDL/RemoteRelationship/Validate.hs

```diff
@@ -405,19 +405,25 @@ validateType permittedVariables value expectedGType types =
                         (G.toGT $ G.NamedType name)
                         "not an input object type"))
 
-assertType :: G.GType -> G.GType -> Validation (NonEmpty ValidationError) ()
-assertType actualType expectedType = do
-  -- check if both are list types or both are named types
-  (when
-     (isListType' actualType /= isListType' expectedType)
-     (Failure (pure $ ExpectedTypeButGot expectedType actualType)))
-  -- if list type then check over unwrapped type, else check base types
-  if isListType' actualType
-    then assertType (unwrapTy actualType) (unwrapTy expectedType)
-    else (when
-            (getBaseTy actualType /= getBaseTy expectedType)
-            (Failure (pure $ ExpectedTypeButGot expectedType actualType)))
-  pure ()
+isTypeCoercible :: G.GType -> G.GType -> Validation (NonEmpty ValidationError) ()
+isTypeCoercible actualType expectedType =
+  -- The GraphQL spec says that, a singleton type can be coerced into  an array
+  -- type. Which means that if the 'actualType' is a singleton type, like
+  -- 'Int' we should be able to join this with a remote node, which expects an
+  -- input argument of type '[Int]'
+  -- http://spec.graphql.org/June2018/#sec-Type-System.List
+  let (actualBaseType, actualNestingLevel) = getBaseTyWithNestedLevelsCount actualType
+      (expectedBaseType, expectedNestingLevel) = getBaseTyWithNestedLevelsCount expectedType
+  in
+  if actualBaseType == expectedBaseType
+  then if (actualNestingLevel == expectedNestingLevel || actualNestingLevel == 0)
+  -- The check of 'actualNestedCount == 0' is the case of coercing a singleton type
+  -- into an array type
+       then pure ()
+       else Failure (pure $ ExpectedTypeButGot expectedType actualType)
+       --  we cannot coerce two types with different nesting levels,
+       -- for example, we cannot coerce [Int] to [[Int]]
+  else Failure (pure $ ExpectedTypeButGot expectedType actualType)
```

Comment:
Can we simplify this code by using MultiwayIf
```haskell
if | actualBaseType /= expectedBaseType -> raiseValidationError
   | actualNestingLevel /= expectedNestingLevel && actualNestingLevel /= 0 -> raiseValidationError
   | otherwise -> pure ()
   where
     raiseValidationError = Failure $ pure $ ExpectedTypeButGot expectedType actualType
```
Of course with appropriate comments.